test/compilable/ob1.d

```diff
@@ -107,3 +107,43 @@ void deallocate(int* ptr, size_t length) @live
 }
 
 
+/*******************************/
+
+void zoo1(int);
+
+@live void zoo2() {
+    int* p = malloc();
+    zoo1(*p);  // does not consume p
+    free(p);
+}
+
+@live void zoo3() {
+    int** p = cast(int**)malloc();
+    free(*p); // consumes p
+}
+
+@live void zoo4() {
+    int[] a = malloc()[0 .. 1];
+    zoo1(a[0]);  // does not consume a
+    free(a.ptr); // consumes a
+}
+
+@live void zoo5() {
+    int*[] a = (cast(int**)malloc())[0 .. 1];
+    free(a[0]); // consumes a
+}
+
+struct S { int i; int* p; }
+
+@live void zoo6() {
+    S* s = cast(S*)malloc();
+    zoo1(s.i);    // does not consume s
+    free(cast(int*)s);
+}
+
+@live void zoo7() {
+    S* s = cast(S*)malloc();
+    free(s.p);    // consumes s
```

Comment:
What's the goal here ? It seems `s` is unusable after one of its member is freed ?
I tried it out locally, did:
```D
@live void zoo7() {
    S* s = cast(S*)malloc(S.sizeof);
    free(s.p);    // consumes s
    s.p = null;
    assert(s.i == 0);
}
```

The two lines I added errored out:
```
ob3.d(16): Error: variable ob3.zoo7.s has undefined state and cannot be read
ob3.d(17): Error: variable ob3.zoo7.s has undefined state and cannot be read
```

This will be problematic for many use cases, for example it's common to have state variables (or handles) that get passed to C functions but are stored as part of other structures.

Reply:
Also tried the following:
```
void foobar (const(int)* ptr);
@live void zoo7() {
    S* s = cast(S*)malloc(S.sizeof);
    foobar(s.p);
    assert(s.i == 0);
}
```

It errors out saying `s` has undefined state.

Reply:
> It errors out saying s has undefined state.

That's right, because you transferred ownership to foobar(). Make the parameter to foobar() `scope`.

Reply:
If it did not work this way, the O/B rules of:

1. one mutable pointer

    OR

2. many constant pointers

   BUT NOT BOTH

would not be valid. It's why using O/B requires redesigning algorithms and data structures.

Reply:
Right, forgot about the `scope` interaction. Will test some more with this in mind.

Reply:
So with this in mind, how are we supposed to `free` fields ?
E.g. the following code causes errors:
```D
@live void zoo7() {
    S* s = cast(S*)malloc(S.sizeof);
    s.p = cast(int*)malloc(int.sizeof * 10);
    foobar(s.p);
    assert(s.i == 0);
    free(s.p);
    free(s); // Double error here
}
```

The double error is:
```
ob3.d(21): Error: variable ob3.zoo7.s has undefined state and cannot be read
ob3.d(21): Error: variable ob3.zoo7.s is not Owner, cannot consume its value
```

Reply:
How, indeed. Welcome to the wonderful world of O/B of redesigning your algorithms and data  structures! It's why people complain a lot about "fighting the borrow checker". The trick is to encapsulate such things and put them in non-@live functions. (This is what Rust does.)

The payoff is memory safety.

Otherwise, suppose `free(s)` is the destructor call for `s`. The destructor naturally frees `s.p`. But you already freed it! Oops!

It is really crucial to understand this aspect of O/B.